### ISSUE TITLE
Codebase Architecture and Code Quality Improvements

### ISSUE THREAD

**Issue by Klinjin** [Klinjin/DH_profile_kSZ_WL#1]
# Codebase Architecture and Code Quality Improvements

## Issue Summary

This issue tracks a comprehensive plan to improve the CAMELS cosmological simulation analysis codebase, focusing on code quality, architecture, scientific computing best practices, and workflow optimization. The improvements are prioritized to maintain the scientific integrity of the project while enhancing maintainability and performance.

### ✅ Completed Components
- **Observation reproduction**: Map stacker and filters working 
- **Basic GP emulation**: Training pipeline functional
- **NPE framework**: SBI integration established

### ⚠️ Issues Requiring Attention  
- **GP accuracy**: Too low for reliable downstream inference
- **Data comparison**: kSZ unit conversion debugging needed, WL needs separate modules
- **NPE reliability**: Corner plots show poor parameter recovery
- **FMPE performance**: Slow sampling from posterior

### ❌ Missing Components
- **Integrated pipeline**: End-to-end cosmology inference workflow
- **Code modularity**: Monolithic functions, duplicate code
- **Testing framework**: No unit tests or validation
- **Configuration management**: Hard-coded paths throughout

#### 1. Eliminate Code Duplication
**Files**: \`GP_dataloader.py:57-95\` 
**Issue**: \`getSims()\` and \`getProfiles()\` contain ~90% duplicate code
**Solution**: 
\`\`\`python
def load_simulation_data(sim_indices, filterType='CAP', ptype='gas', 
                        include_params=True, include_pk=True, include_mass=True):
    \"\"\"Unified data loading function with configurable outputs.\"\"\"
\`\`\`
**Impact**: Reduces maintenance burden, eliminates inconsistencies

#### 2. Centralize Configuration  
**Files**: \`GP_dataloader.py:16,64,125,177\`, \`train_GP.py:28-32\`
**Issue**: Hard-coded paths scattered throughout codebase
**Solution**: Create \`config.py\`:
\`\`\`python
# Configuration constants
CAMELS_BASE_PATH = '/pscratch/sd/l/lindajin/CAMELS/IllustrisTNG/L50n512_SB35'
PROFILE_FILENAME = 'Henry_profiles_gas_dm_star_bh_nPixel1000_R_lin0.04_2.5_log15_nbins20.npz'
PARAM_MATRIX_PATH = 'data/camels_params_matrix.npy'
\`\`\`
**Impact**: Single point of configuration, easier deployment

#### 3. Remove Duplicate Imports
**Files**: \`train_GP.py:16-17\`, \`train_GP.py:28-32\`  
**Issue**: Multiple JAX imports, redundant environment setup
**Solution**: Create \`src/utils/environment.py\` for shared setup
**Impact**: Cleaner imports, consistent environment

#### 4. Restructure Directory Layout
**Current**: Flat structure with mixed concerns  
**Proposed**:
\`\`\`
src/
├── data/           # GP_dataloader.py → data_loader.py
├── models/         # train_GP.py → gp_trainer.py, neural_gp.py  
├── inference/      # NPE.py → sbi_inference.py
├── utils/          # filter_utils.py, stacker.py, etc.
├── config/         # config.py, constants.py
└── experiments/    # Move notebooks here
tests/
├── test_data_loader.py
├── test_gp_training.py
└── test_npe_inference.py
\`\`\`
**Impact**: Better separation of concerns, easier navigation

#### 5. Decompose Monolithic Functions
**File**: \`train_GP.py:86-179\` (\`train_conditional_gp\`)
**Issue**: Single function handles data prep, training, optimization, and saving
**Solution**: Split into focused functions:
- \`prepare_training_data()\` 
- \`train_single_gp_model()\`
- \`optimize_hyperparameters()\`
- \`save_trained_models()\`
**Impact**: Better testability, reusability, debugging

#### 6. Add Experiment Reproducibility
**File**: \`NPE.py:43-50\`
**Issue**: No experiment tracking, random seeds not managed
**Solution**: Configuration-driven experiments:
\`\`\`yaml
# experiments/configs/npe_baseline.yaml
experiment:
  name: "camels_npe_baseline_v1"
  random_seed: 12345
  method: "NPE"
  num_sims: 1000
  sim_id: 777
  num_rounds: 3
\`\`\`
**Impact**: Reproducible results, experiment comparison

#### 7. Add Data Validation  
**File**: \`GP_dataloader.py:69-70\`
**Issue**: Silent failures from data transformation issues
**Solution**: Add validation functions:
\`\`\`python
def validate_profiles(profiles, name="profiles"):
    if np.any(np.isnan(profiles)):
        raise ValueError(f"NaN values found in {name}")
    if np.any(profiles <= 0):
        warnings.warn(f"Non-positive values in {name}")
    return profiles
\`\`\`
**Impact**: Catch data issues early, prevent silent failures

#### 8. Implement Efficient Data Loading
**File**: \`GP_dataloader.py:112-121\`
**Issue**: O(n²) parameter matching in \`getProfilesParamsTensor()\`
**Solution**: Pre-compute KDTree for nearest neighbor search
**Impact**: Faster parameter lookup, especially for large datasets

#### 9. Memory-Efficient Data Loading
**File**: \`train_GP.py:61-83\` (\`prepare_GP_data\`)  
**Issue**: Loads all simulation data into memory at once
**Solution**: Implement PyTorch Dataset/DataLoader pattern:
\`\`\`python
class SimulationDataset(torch.utils.data.Dataset):
    def __getitem__(self, idx):
        # Load only requested simulation on demand
\`\`\`
**Impact**: Reduced memory usage, scalability to larger datasets

#### 10. Add Comprehensive Testing
**Missing**: No unit tests exist
**Solution**: Create \`tests/\` directory with:
- \`test_data_loader.py\`: Test all GP_dataloader functions
- \`test_gp_training.py\`: Test GP model training and validation  
- \`test_npe_inference.py\`: Test SBI inference pipeline
- \`test_utils.py\`: Test utility functions
**Impact**: Catch regressions, enable confident refactoring

#### 11. Add Structured Logging
**All Files**: Print statements instead of proper logging
**Solution**: Create \`src/utils/logger.py\`:
\`\`\`python
import logging
logger = logging.getLogger('camels_analysis')
# Configure structured logging with timestamps, levels
\`\`\`
**Impact**: Better debugging, production monitoring

#### 12. Improve Documentation  
**Files**: Most functions lack comprehensive docstrings
**Priority locations**:
- \`GP_dataloader.py:6\` (\`getSims\`) - add parameter descriptions
- \`train_GP.py:182\` (\`build_hierarchical_gp\`) - explain kernel construction  
- \`NPE.py:47\` (\`simulator\`) - document input/output format
**Impact**: Better code understanding, easier onboarding

### Critical Issue: GP Accuracy 
**File**: \`train_GP.py:106-114\`
**Problem**: Fixed hyperparameters lead to poor GP predictions
\`\`\`python
# Current - too simplistic
params = {
    "cosmo_amplitude": jnp.float32(0.0),
    "cosmo_length_scales": jnp.zeros(n_cosmo_params),  # All same scale
}
\`\`\`
**Solution**: Implement hyperparameter optimization
**Impact**: ⚠️ **This is blocking NPE performance** - fix first priority

#### 1. kSZ Unit Validation
**File**: \`ksz_utils.py\` (needs examination)  
**Issue**: Unit conversion debugging required vs DESI data
**Solution**: Create \`src/analysis/ksz_validator.py\`

#### 2. FMPE Performance  
**File**: \`NPE.py:75-79\`
**Issue**: Slow posterior sampling  
**Solution**: 
- Implement batch sampling
- GPU acceleration 
- Consider MCMC alternatives

#### 3. End-to-End Pipeline
**Missing**: Integrated observation → cosmology inference
**Solution**: Create \`src/pipeline/inference_pipeline.py\`:
\`\`\`python
class CosmologyInferencePipeline:
    def infer_cosmology(self, observation_data, obs_type='kSZ'):
        # observation → GP prediction → NPE inference → results
\`\`\`

### Week 1: Foundation (High Impact)
1. **Create \`config.py\`** - centralize hard-coded paths
2. **Refactor \`GP_dataloader.py\`** - eliminate duplication  
3. **Fix GP hyperparameters** - improve accuracy

### Week 2: Architecture  
1. **Restructure directories** - implement \`src/\` layout
2. **Split \`train_GP.py\`** - decompose monolithic functions
3. **Add data validation** - prevent silent failures

### Week 3: Scientific Improvements
1. **Optimize GP training** - hyperparameter tuning
2. **Debug kSZ units** - validate against DESI  
3. **Add experiment tracking** - reproducible NPE runs

### Week 4: Integration & Testing
1. **Create end-to-end pipeline** - observation to inference
2. **Add unit tests** - ensure reliability
3. **Performance optimization** - FMPE acceleration

## Success Metrics

- [x] **Code Quality**: No duplicate code, centralized config
- [x] **GP Performance**: Improved accuracy metrics, validated predictions  
- [ ] **NPE Reliability**: Corner plots show accurate parameter recovery
- [ ] **Pipeline Integration**: End-to-end cosmology inference working
- [ ] **Scientific Validation**: kSZ units verified against DESI data
- [ ] **Testing Coverage**: All core functions have unit tests

### External Dependencies
- JAX/TinyGP for GP models (no changes needed)
- SBI library for NPE (may need version compatibility check)
- Simulation data access (paths need to remain functional)

### Migration Risks  
- **Data path changes**: Ensure backward compatibility during config migration
- **Function signature changes**: May break existing notebooks
- **Import path changes**: Will require updating notebook imports

### Mitigation Strategy
1. **Phased migration**: Implement alongside existing code initially
2. **Backward compatibility**: Maintain old function names as deprecated wrappers  
3. **Validation testing**: Compare outputs before/after refactoring

---

## Next Steps

**Immediate priorities** (this week):
1. Fix GP accuracy issue (blocking downstream performance)
2. Create configuration system  
3. Eliminate code duplication in data loading

**Medium term** (next 2 weeks):
1. Restructure codebase architecture
2. Add comprehensive testing
3. Debug kSZ unit conversions

**Long term** (next month):  
1. Integrated inference pipeline
2. Performance optimizations
3. Scientific validation against DESI data

The improvements maintain scientific integrity while significantly enhancing code maintainability, performance, and reliability. The GP accuracy fix is the highest priority as it's currently blocking reliable parameter inference.
